File: test/reference-output/generics.ts
import * as svt from "simple-validation-tools";

import * as basic from "./basic";

import * as hasGeneric from "./hasGeneric";

export type UsingGenerics = {
    field1: basic.Maybe<string>;
    field2: basic.Either<string, number>;
};

export function isUsingGenerics(value: unknown): value is UsingGenerics {
    return svt.isInterface<UsingGenerics>(value, {field1: basic.isMaybe(svt.isString), field2: basic.isEither(svt.isString, svt.isNumber)});
}

export function validateUsingGenerics(value: unknown): svt.ValidationResult<UsingGenerics> {
    return svt.validate<UsingGenerics>(value, {field1: basic.validateMaybe(svt.validateString), field2: basic.validateEither(svt.validateString, svt.validateNumber)});
}

export type UsingOwnGenerics<T> = {
    field1: basic.Maybe<T>;
};

export function isUsingOwnGenerics<T>(isT: svt.TypePredicate<T>): svt.TypePredicate<UsingOwnGenerics<T>> {
    return function isUsingOwnGenericsT(value: unknown): value is UsingOwnGenerics<T> {
        return svt.isInterface<UsingOwnGenerics<T>>(value, {field1: basic.isMaybe(isT)});
    };
}

export function validateUsingOwnGenerics<T>(validateT: svt.Validator<T>): svt.Validator<UsingOwnGenerics<T>> {
    return function validateUsingOwnGenericsT(value: unknown): svt.ValidationResult<UsingOwnGenerics<T>> {
        return svt.validate<UsingOwnGenerics<T>>(value, {field1: basic.validateMaybe(validateT)});
    };
}

export type KnownForMovie = {
    media_type: "movie";
    poster_path: string | null | undefined;
    id: number;
    title: string | null | undefined;
    vote_average: number;
    release_date: string | null | undefined;
    overview: string;
};

export function isKnownForMovie(value: unknown): value is KnownForMovie {
    return svt.isInterface<KnownForMovie>(value, {media_type: "movie", poster_path: svt.optional(svt.isString), id: svt.isNumber, title: svt.optional(svt.isString), vote_average: svt.isNumber, release_date: svt.optional(svt.isString), overview: svt.isString});
}

export function validateKnownForMovie(value: unknown): svt.ValidationResult<KnownForMovie> {
    return svt.validate<KnownForMovie>(value, {media_type: "movie", poster_path: svt.validateOptional(svt.validateString), id: svt.validateNumber, title: svt.validateOptional(svt.validateString), vote_average: svt.validateNumber, release_date: svt.validateOptional(svt.validateString), overview: svt.validateString});
}

export type KnownForShow = {
    media_type: "tv";
    poster_path: string | null | undefined;
    id: number;
    vote_average: number;
    overview: string;
    first_air_date: string | null | undefined;
    name: string | null | undefined;
};

export function isKnownForShow(value: unknown): value is KnownForShow {
    return svt.isInterface<KnownForShow>(value, {media_type: "tv", poster_path: svt.optional(svt.isString), id: svt.isNumber, vote_average: svt.isNumber, overview: svt.isString, first_air_date: svt.optional(svt.isString), name: svt.optional(svt.isString)});
}

export function validateKnownForShow(value: unknown): svt.ValidationResult<KnownForShow> {
    return svt.validate<KnownForShow>(value, {media_type: "tv", poster_path: svt.validateOptional(svt.validateString), id: svt.validateNumber, vote_average: svt.validateNumber, overview: svt.validateString, first_air_date: svt.validateOptional(svt.validateString), name: svt.validateOptional(svt.validateString)});
}

export type KnownFor = KnownForShow | KnownForMovie | string | number;

export function isKnownFor(value: unknown): value is KnownFor {
    return [isKnownForShow, isKnownForMovie, svt.isString, svt.isNumber].some((typePredicate) => typePredicate(value));
}

export function validateKnownFor(value: unknown): svt.ValidationResult<KnownFor> {
    return svt.validateOneOf<KnownFor>(value, [validateKnownForShow, validateKnownForMovie, svt.validateString, svt.validateNumber]);
}

export type KnownForMovieWithoutTypeTag = {
    poster_path: string | null | undefined;
    id: number;
    title: string | null | undefined;
    vote_average: number;
    release_date: string | null | undefined;
    overview: string;
};

export function isKnownForMovieWithoutTypeTag(value: unknown): value is KnownForMovieWithoutTypeTag {
    return svt.isInterface<KnownForMovieWithoutTypeTag>(value, {poster_path: svt.optional(svt.isString), id: svt.isNumber, title: svt.optional(svt.isString), vote_average: svt.isNumber, release_date: svt.optional(svt.isString), overview: svt.isString});
}

export function validateKnownForMovieWithoutTypeTag(value: unknown): svt.ValidationResult<KnownForMovieWithoutTypeTag> {
    return svt.validate<KnownForMovieWithoutTypeTag>(value, {poster_path: svt.validateOptional(svt.validateString), id: svt.validateNumber, title: svt.validateOptional(svt.validateString), vote_average: svt.validateNumber, release_date: svt.validateOptional(svt.validateString), overview: svt.validateString});
}

export type KnownForShowWithoutTypeTag = {
    poster_path: string | null | undefined;
    id: number;
    vote_average: number;
    overview: string;
    first_air_date: string | null | undefined;
    name: string | null | undefined;
};

export function isKnownForShowWithoutTypeTag(value: unknown): value is KnownForShowWithoutTypeTag {
    return svt.isInterface<KnownForShowWithoutTypeTag>(value, {poster_path: svt.optional(svt.isString), id: svt.isNumber, vote_average: svt.isNumber, overview: svt.isString, first_air_date: svt.optional(svt.isString), name: svt.optional(svt.isString)});
}

export function validateKnownForShowWithoutTypeTag(value: unknown): svt.ValidationResult<KnownForShowWithoutTypeTag> {
    return svt.validate<KnownForShowWithoutTypeTag>(value, {poster_path: svt.validateOptional(svt.validateString), id: svt.validateNumber, vote_average: svt.validateNumber, overview: svt.validateString, first_air_date: svt.validateOptional(svt.validateString), name: svt.validateOptional(svt.validateString)});
}

export type KnownForEmbedded = MovieStartingWithLowercase | TvStartingWithLowercase;

export enum KnownForEmbeddedTag {
    MovieStartingWithLowercase = "movieStartingWithLowercase",
    TvStartingWithLowercase = "tvStartingWithLowercase",
}

export type MovieStartingWithLowercase = {
    media_type: KnownForEmbeddedTag.MovieStartingWithLowercase;
    poster_path: string | null | undefined;
    id: number;
    title: string | null | undefined;
    vote_average: number;
    release_date: string | null | undefined;
    overview: string;
};

export type TvStartingWithLowercase = {
    media_type: KnownForEmbeddedTag.TvStartingWithLowercase;
    poster_path: string | null | undefined;
    id: number;
    vote_average: number;
    overview: string;
    first_air_date: string | null | undefined;
    name: string | null | undefined;
};

export function MovieStartingWithLowercase(data: KnownForMovieWithoutTypeTag): MovieStartingWithLowercase {
    return {media_type: KnownForEmbeddedTag.MovieStartingWithLowercase, ...data};
}

export function TvStartingWithLowercase(data: KnownForShowWithoutTypeTag): TvStartingWithLowercase {
    return {media_type: KnownForEmbeddedTag.TvStartingWithLowercase, ...data};
}

export function isKnownForEmbedded(value: unknown): value is KnownForEmbedded {
    return [isMovieStartingWithLowercase, isTvStartingWithLowercase].some((typePredicate) => typePredicate(value));
}

export function isMovieStartingWithLowercase(value: unknown): value is MovieStartingWithLowercase {
    return svt.isInterface<MovieStartingWithLowercase>(value, {media_type: KnownForEmbeddedTag.MovieStartingWithLowercase, poster_path: svt.optional(svt.isString), id: svt.isNumber, title: svt.optional(svt.isString), vote_average: svt.isNumber, release_date: svt.optional(svt.isString), overview: svt.isString});
}

export function isTvStartingWithLowercase(value: unknown): value is TvStartingWithLowercase {
    return svt.isInterface<TvStartingWithLowercase>(value, {media_type: KnownForEmbeddedTag.TvStartingWithLowercase, poster_path: svt.optional(svt.isString), id: svt.isNumber, vote_average: svt.isNumber, overview: svt.isString, first_air_date: svt.optional(svt.isString), name: svt.optional(svt.isString)});
}

export function validateKnownForEmbedded(value: unknown): svt.ValidationResult<KnownForEmbedded> {
    return svt.validateWithTypeTag<KnownForEmbedded>(value, {[KnownForEmbeddedTag.MovieStartingWithLowercase]: validateMovieStartingWithLowercase, [KnownForEmbeddedTag.TvStartingWithLowercase]: validateTvStartingWithLowercase}, "media_type");
}

export function validateMovieStartingWithLowercase(value: unknown): svt.ValidationResult<MovieStartingWithLowercase> {
    return svt.validate<MovieStartingWithLowercase>(value, {media_type: KnownForEmbeddedTag.MovieStartingWithLowercase, poster_path: svt.validateOptional(svt.validateString), id: svt.validateNumber, title: svt.validateOptional(svt.validateString), vote_average: svt.validateNumber, release_date: svt.validateOptional(svt.validateString), overview: svt.validateString});
}

export function validateTvStartingWithLowercase(value: unknown): svt.ValidationResult<TvStartingWithLowercase> {
    return svt.validate<TvStartingWithLowercase>(value, {media_type: KnownForEmbeddedTag.TvStartingWithLowercase, poster_path: svt.validateOptional(svt.validateString), id: svt.validateNumber, vote_average: svt.validateNumber, overview: svt.validateString, first_air_date: svt.validateOptional(svt.validateString), name: svt.validateOptional(svt.validateString)});
}

export type KnownForEmbeddedWithUpperCase = Movie | Tv;

export enum KnownForEmbeddedWithUpperCaseTag {
    Movie = "Movie",
    Tv = "Tv",
}

export type Movie = {
    media_type: KnownForEmbeddedWithUpperCaseTag.Movie;
    poster_path: string | null | undefined;
    id: number;
    title: string | null | undefined;
    vote_average: number;
    release_date: string | null | undefined;
    overview: string;
};

export type Tv = {
    media_type: KnownForEmbeddedWithUpperCaseTag.Tv;
    poster_path: string | null | undefined;
    id: number;
    vote_average: number;
    overview: string;
    first_air_date: string | null | undefined;
    name: string | null | undefined;
};

export function Movie(data: KnownForMovieWithoutTypeTag): Movie {
    return {media_type: KnownForEmbeddedWithUpperCaseTag.Movie, ...data};
}

export function Tv(data: KnownForShowWithoutTypeTag): Tv {
    return {media_type: KnownForEmbeddedWithUpperCaseTag.Tv, ...data};
}

export function isKnownForEmbeddedWithUpperCase(value: unknown): value is KnownForEmbeddedWithUpperCase {
    return [isMovie, isTv].some((typePredicate) => typePredicate(value));
}

export function isMovie(value: unknown): value is Movie {
    return svt.isInterface<Movie>(value, {media_type: KnownForEmbeddedWithUpperCaseTag.Movie, poster_path: svt.optional(svt.isString), id: svt.isNumber, title: svt.optional(svt.isString), vote_average: svt.isNumber, release_date: svt.optional(svt.isString), overview: svt.isString});
}

export function isTv(value: unknown): value is Tv {
    return svt.isInterface<Tv>(value, {media_type: KnownForEmbeddedWithUpperCaseTag.Tv, poster_path: svt.optional(svt.isString), id: svt.isNumber, vote_average: svt.isNumber, overview: svt.isString, first_air_date: svt.optional(svt.isString), name: svt.optional(svt.isString)});
}

export function validateKnownForEmbeddedWithUpperCase(value: unknown): svt.ValidationResult<KnownForEmbeddedWithUpperCase> {
    return svt.validateWithTypeTag<KnownForEmbeddedWithUpperCase>(value, {[KnownForEmbeddedWithUpperCaseTag.Movie]: validateMovie, [KnownForEmbeddedWithUpperCaseTag.Tv]: validateTv}, "media_type");
}

export function validateMovie(value: unknown): svt.ValidationResult<Movie> {
    return svt.validate<Movie>(value, {media_type: KnownForEmbeddedWithUpperCaseTag.Movie, poster_path: svt.validateOptional(svt.validateString), id: svt.validateNumber, title: svt.validateOptional(svt.validateString), vote_average: svt.validateNumber, release_date: svt.validateOptional(svt.validateString), overview: svt.validateString});
}

export function validateTv(value: unknown): svt.ValidationResult<Tv> {
    return svt.validate<Tv>(value, {media_type: KnownForEmbeddedWithUpperCaseTag.Tv, poster_path: svt.validateOptional(svt.validateString), id: svt.validateNumber, vote_average: svt.validateNumber, overview: svt.validateString, first_air_date: svt.validateOptional(svt.validateString), name: svt.validateOptional(svt.validateString)});
}

File: test/reference-output/generics.py
import enum
import json
import typing
from dataclasses import dataclass
from gotyno_validation import encoding, validation

from . import basic
from . import hasGeneric

@dataclass(frozen=True)
class UsingGenerics:
    field1: basic.Maybe[str]
    field2: basic.Either[str, int]

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['UsingGenerics']:
        return validation.validate_interface(value, {'field1': basic.Maybe.validate(validation.validate_string), 'field2': basic.Either.validate(validation.validate_string, validation.validate_int)}, UsingGenerics)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['UsingGenerics']:
        return validation.validate_from_string(string, UsingGenerics.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'field1': self.field1.to_json(encoding.basic_to_json), 'field2': self.field2.to_json(encoding.basic_to_json, encoding.basic_to_json)}

    def encode(self) -> str:
        return json.dumps(self.to_json())

T = typing.TypeVar('T')
@dataclass(frozen=True)
class UsingOwnGenerics(typing.Generic[T]):
    field1: basic.Maybe[T]

    @staticmethod
    def validate(validate_T: validation.Validator[T]) -> validation.Validator['UsingOwnGenerics[T]']:
        def validate_UsingOwnGenericsT(value: validation.Unknown) -> validation.ValidationResult['UsingOwnGenerics[T]']:
            return validation.validate_interface(value, {'field1': basic.Maybe.validate(validate_T)}, UsingOwnGenerics)
        return validate_UsingOwnGenericsT

    @staticmethod
    def decode(string: typing.Union[str, bytes], validate_T: validation.Validator[T]) -> validation.ValidationResult['UsingOwnGenerics[T]']:
        return validation.validate_from_string(string, UsingOwnGenerics.validate(validate_T))

    def to_json(self, T_to_json: encoding.ToJSON[T]) -> typing.Dict[str, typing.Any]:
        return {'field1': self.field1.to_json(T_to_json)}

    def encode(self, T_to_json: encoding.ToJSON[T]) -> str:
        return json.dumps(self.to_json(T_to_json))

@dataclass(frozen=True)
class KnownForMovie:
    media_type: typing.Literal['movie']
    poster_path: typing.Optional[str]
    id: int
    title: typing.Optional[str]
    vote_average: float
    release_date: typing.Optional[str]
    overview: str

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['KnownForMovie']:
        return validation.validate_interface(value, {'media_type': validation.validate_literal('movie'), 'poster_path': validation.validate_optional(validation.validate_string), 'id': validation.validate_int, 'title': validation.validate_optional(validation.validate_string), 'vote_average': validation.validate_float, 'release_date': validation.validate_optional(validation.validate_string), 'overview': validation.validate_string}, KnownForMovie)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['KnownForMovie']:
        return validation.validate_from_string(string, KnownForMovie.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'media_type': 'movie', 'poster_path': encoding.optional_to_json(encoding.basic_to_json)(self.poster_path), 'id': self.id, 'title': encoding.optional_to_json(encoding.basic_to_json)(self.title), 'vote_average': self.vote_average, 'release_date': encoding.optional_to_json(encoding.basic_to_json)(self.release_date), 'overview': self.overview}

    def encode(self) -> str:
        return json.dumps(self.to_json())

@dataclass(frozen=True)
class KnownForShow:
    media_type: typing.Literal['tv']
    poster_path: typing.Optional[str]
    id: int
    vote_average: float
    overview: str
    first_air_date: typing.Optional[str]
    name: typing.Optional[str]

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['KnownForShow']:
        return validation.validate_interface(value, {'media_type': validation.validate_literal('tv'), 'poster_path': validation.validate_optional(validation.validate_string), 'id': validation.validate_int, 'vote_average': validation.validate_float, 'overview': validation.validate_string, 'first_air_date': validation.validate_optional(validation.validate_string), 'name': validation.validate_optional(validation.validate_string)}, KnownForShow)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['KnownForShow']:
        return validation.validate_from_string(string, KnownForShow.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'media_type': 'tv', 'poster_path': encoding.optional_to_json(encoding.basic_to_json)(self.poster_path), 'id': self.id, 'vote_average': self.vote_average, 'overview': self.overview, 'first_air_date': encoding.optional_to_json(encoding.basic_to_json)(self.first_air_date), 'name': encoding.optional_to_json(encoding.basic_to_json)(self.name)}

    def encode(self) -> str:
        return json.dumps(self.to_json())

KnownFor = typing.Union[KnownForShow, KnownForMovie, str, float]
class KnownForInterface:
    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['KnownFor']:
        return validation.validate_one_of(value, [KnownForShow.validate, KnownForMovie.validate, validation.validate_string, validation.validate_float])

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['KnownFor']:
        return validation.validate_from_string(string, KnownForInterface.validate)

    @staticmethod
    def to_json(value) -> typing.Any:
        return encoding.one_of_to_json(value, {KnownForShow: KnownForShow.to_json, KnownForMovie: KnownForMovie.to_json, str: encoding.basic_to_json, float: encoding.basic_to_json})

    @staticmethod
    def encode(value) -> str:
        return json.dumps(value.to_json())

@dataclass(frozen=True)
class KnownForMovieWithoutTypeTag:
    poster_path: typing.Optional[str]
    id: int
    title: typing.Optional[str]
    vote_average: float
    release_date: typing.Optional[str]
    overview: str

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['KnownForMovieWithoutTypeTag']:
        return validation.validate_interface(value, {'poster_path': validation.validate_optional(validation.validate_string), 'id': validation.validate_int, 'title': validation.validate_optional(validation.validate_string), 'vote_average': validation.validate_float, 'release_date': validation.validate_optional(validation.validate_string), 'overview': validation.validate_string}, KnownForMovieWithoutTypeTag)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['KnownForMovieWithoutTypeTag']:
        return validation.validate_from_string(string, KnownForMovieWithoutTypeTag.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'poster_path': encoding.optional_to_json(encoding.basic_to_json)(self.poster_path), 'id': self.id, 'title': encoding.optional_to_json(encoding.basic_to_json)(self.title), 'vote_average': self.vote_average, 'release_date': encoding.optional_to_json(encoding.basic_to_json)(self.release_date), 'overview': self.overview}

    def encode(self) -> str:
        return json.dumps(self.to_json())

@dataclass(frozen=True)
class KnownForShowWithoutTypeTag:
    poster_path: typing.Optional[str]
    id: int
    vote_average: float
    overview: str
    first_air_date: typing.Optional[str]
    name: typing.Optional[str]

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['KnownForShowWithoutTypeTag']:
        return validation.validate_interface(value, {'poster_path': validation.validate_optional(validation.validate_string), 'id': validation.validate_int, 'vote_average': validation.validate_float, 'overview': validation.validate_string, 'first_air_date': validation.validate_optional(validation.validate_string), 'name': validation.validate_optional(validation.validate_string)}, KnownForShowWithoutTypeTag)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['KnownForShowWithoutTypeTag']:
        return validation.validate_from_string(string, KnownForShowWithoutTypeTag.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'poster_path': encoding.optional_to_json(encoding.basic_to_json)(self.poster_path), 'id': self.id, 'vote_average': self.vote_average, 'overview': self.overview, 'first_air_date': encoding.optional_to_json(encoding.basic_to_json)(self.first_air_date), 'name': encoding.optional_to_json(encoding.basic_to_json)(self.name)}

    def encode(self) -> str:
        return json.dumps(self.to_json())

class KnownForEmbedded:
    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['KnownForEmbedded']:
        return validation.validate_with_type_tags(value, 'media_type', {'movieStartingWithLowercase': MovieStartingWithLowercase.validate, 'tvStartingWithLowercase': TvStartingWithLowercase.validate})

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['KnownForEmbedded']:
        return validation.validate_from_string(string, KnownForEmbedded.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        raise NotImplementedError('`to_json` is not implemented for base class `KnownForEmbedded`')

    def encode(self) -> str:
        raise NotImplementedError('`encode` is not implemented for base class `KnownForEmbedded`')

@dataclass
class MovieStartingWithLowercase(KnownForEmbedded):
    poster_path: typing.Optional[str]
    id: int
    title: typing.Optional[str]
    vote_average: float
    release_date: typing.Optional[str]
    overview: str

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['MovieStartingWithLowercase']:
        return validation.validate_with_type_tag_and_validator(value, 'media_type', 'movieStartingWithLowercase', KnownForMovieWithoutTypeTag.validate, MovieStartingWithLowercase)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['MovieStartingWithLowercase']:
        return validation.validate_from_string(string, MovieStartingWithLowercase.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'media_type': 'movieStartingWithLowercase', **KnownForMovieWithoutTypeTag.to_json(self)}

    def encode(self) -> str:
        return json.dumps(self.to_json())

@dataclass
class TvStartingWithLowercase(KnownForEmbedded):
    poster_path: typing.Optional[str]
    id: int
    vote_average: float
    overview: str
    first_air_date: typing.Optional[str]
    name: typing.Optional[str]

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['TvStartingWithLowercase']:
        return validation.validate_with_type_tag_and_validator(value, 'media_type', 'tvStartingWithLowercase', KnownForShowWithoutTypeTag.validate, TvStartingWithLowercase)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['TvStartingWithLowercase']:
        return validation.validate_from_string(string, TvStartingWithLowercase.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'media_type': 'tvStartingWithLowercase', **KnownForShowWithoutTypeTag.to_json(self)}

    def encode(self) -> str:
        return json.dumps(self.to_json())

class KnownForEmbeddedWithUpperCase:
    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['KnownForEmbeddedWithUpperCase']:
        return validation.validate_with_type_tags(value, 'media_type', {'Movie': Movie.validate, 'Tv': Tv.validate})

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['KnownForEmbeddedWithUpperCase']:
        return validation.validate_from_string(string, KnownForEmbeddedWithUpperCase.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        raise NotImplementedError('`to_json` is not implemented for base class `KnownForEmbeddedWithUpperCase`')

    def encode(self) -> str:
        raise NotImplementedError('`encode` is not implemented for base class `KnownForEmbeddedWithUpperCase`')

@dataclass
class Movie(KnownForEmbeddedWithUpperCase):
    poster_path: typing.Optional[str]
    id: int
    title: typing.Optional[str]
    vote_average: float
    release_date: typing.Optional[str]
    overview: str

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['Movie']:
        return validation.validate_with_type_tag_and_validator(value, 'media_type', 'Movie', KnownForMovieWithoutTypeTag.validate, Movie)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['Movie']:
        return validation.validate_from_string(string, Movie.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'media_type': 'Movie', **KnownForMovieWithoutTypeTag.to_json(self)}

    def encode(self) -> str:
        return json.dumps(self.to_json())

@dataclass
class Tv(KnownForEmbeddedWithUpperCase):
    poster_path: typing.Optional[str]
    id: int
    vote_average: float
    overview: str
    first_air_date: typing.Optional[str]
    name: typing.Optional[str]

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['Tv']:
        return validation.validate_with_type_tag_and_validator(value, 'media_type', 'Tv', KnownForShowWithoutTypeTag.validate, Tv)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['Tv']:
        return validation.validate_from_string(string, Tv.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'media_type': 'Tv', **KnownForShowWithoutTypeTag.to_json(self)}

    def encode(self) -> str:
        return json.dumps(self.to_json())